### PR TITLE
Configurable file descriptor limit

### PR DESCRIPTION
**Pull request by Lignum**
This adds a configurable file handle limit, defaulting to 128, to prevent players from exhausting the server's file descriptors. This is in response to issue #144.

The limit is always active and can be configured with `I:maximumFilesOpen`. To "disable" it, just set the limit to a really high value. Negative or zero values will simply prevent files from being read and written, making all computers useless.

Once the limit is reached, `fs.open` will return `nil`.

### CODE REVIEW

File: src/main/java/dan200/computercraft/core/filesystem/FileSystem.java [hunk 1]
```diff
@@ -649,6 +650,33 @@ private synchronized void copyRecursive( String sourcePath, MountWrapper sourceM
...
         }
     }
+
+    private synchronized IMountedFile openFile( IMountedFile file ) throws FileSystemException
+    {
+        synchronized( m_openFiles )
+        {
+            if( m_openFiles.size() >= ComputerCraft.maximumFilesOpen )
+            {
+                throw new FileSystemException("Too many files already open");
```
Comment (SquidDev) [inline src/main/java/dan200/computercraft/core/filesystem/FileSystem.java]: You're not closing the file handle here.

**Comment by SquidDev** [comment]
Heh, this was next on my todo list - yah beat me to it 😛. 

One issue with this implementation is that `FileSystemException` is caught by `fs.open`. It just returns `null` instead, swallowing the error. You'd probably have to create a separate exception, or throw a `LuaException` instead. Neither solution seems especially elegant.

**Comment by Lignum** [comment]
I think that's intentional? Since there's a "No such file" exception being thrown when you open a file that doesn't exist, but `fs.open` also just returns `nil`.

File: src/main/java/dan200/computercraft/core/filesystem/FileSystem.java (hunk 1 above)
Comment (Lignum) [inline src/main/java/dan200/computercraft/core/filesystem/FileSystem.java]: Oops, will fix in a second.

File: src/main/java/dan200/computercraft/core/filesystem/FileSystem.java [hunk 2]
```diff
@@ -697,11 +729,7 @@ public void flush() throws IOException
                     throw new UnsupportedOperationException();
                 }
             };
-            synchronized( m_openFiles )
-            {
-                m_openFiles.add( file );
-            }
-            return file;
+            return (IMountedFileNormal) openFile( file, reader );
```
Comment (SquidDev) [inline src/main/java/dan200/computercraft/core/filesystem/FileSystem.java]: Do you need the cast here? Can you not just do:

```java
openFile( file, reader );
return file;
```

Comment (Lignum) [inline src/main/java/dan200/computercraft/core/filesystem/FileSystem.java]: I can, but I wanted a one-liner, which is why `openFile` returns the handle in the first place. So unless that cast is bad for some reason, it's down to preference, I guess.

Comment (SquidDev) [inline src/main/java/dan200/computercraft/core/filesystem/FileSystem.java]: I generally don't like casts unless you need them. I guess you could make `openFile` generic:

```java
private synchronized <T extends IMountedFile> T openFile( T file, Closeable handle ) throws FileSystemException
```

Though I'm not sure that's any nicer 😄.

Comment (JLLeitschuh) [inline src/main/java/dan200/computercraft/core/filesystem/FileSystem.java]:
```
private synchronized <T extends IMountedFile> T openFile( T file, Closeable handle ) throws FileSystemException
```
I think that would accurately capture what is trying to be done here.
It actually seems like a better solution.

Comment (Lignum) [inline src/main/java/dan200/computercraft/core/filesystem/FileSystem.java]: I've gone with that solution now.

**Comment by dan200** [comment]
This limit is per-computer, not per-server. Is that the intent?

**Comment by Lignum** [comment]
Yes, that's the intent.

**Comment by Lignum** [comment]
Cheers! :tada:

**Pull request merged**